core/drivers/crypto/hisilicon/sec/sec_cipher.c

```diff
@@ -0,0 +1,781 @@
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_alloc_buffer(struct sec_cipher_ctx *c_ctx)
+{
+	c_ctx->in = (uint8_t *)malloc(c_ctx->len);
+	if (!c_ctx->in) {
+		EMSG("Failed to alloc c_in buf");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	c_ctx->in_dma = virt_to_phys(c_ctx->in);
+	assert(c_ctx->in_dma);
+
+	c_ctx->out = (uint8_t *)malloc(c_ctx->len);
+	if (!c_ctx->out) {
+		EMSG("Failed to alloc c_out buf");
+		goto free_c_in;
+	}
+
+	c_ctx->out_dma = virt_to_phys(c_ctx->out);
+	assert(c_ctx->out_dma);
+
+	return TEE_SUCCESS;
+
+free_c_in:
+	free(c_ctx->in);
+	return TEE_ERROR_STORAGE_NO_SPACE;
+}
+
+static void sec_free_buffer(struct sec_cipher_ctx *c_ctx)
+{
+	free(c_ctx->in);
+	free(c_ctx->out);
+}
+
+static TEE_Result sec_cipher_update(struct drvcrypt_cipher_update *dupdate)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	size_t padding_size = 0;
+	TEE_Result ret = 0;
+
+	ret = sec_cipher_param_check(dupdate);
+	if (ret != 0)
+		return ret;
+
+	c_ctx = (struct sec_cipher_ctx *)dupdate->ctx;
+	if (c_ctx->mode == C_MODE_CTR && (c_ctx->offs & CTR_SRC_ALIGN_MASK))
+		padding_size = (c_ctx->offs % CTR_SRC_BLOCK_SIZE);
+
+	c_ctx->offs += dupdate->src.length;
+	c_ctx->len = dupdate->src.length + padding_size;
+	ret = sec_alloc_buffer(c_ctx);
+	if (ret)
+		return ret;
+
+	memset(c_ctx->in, 0, padding_size);
+	memcpy(c_ctx->in + padding_size,
+	       dupdate->src.data, dupdate->src.length);
+
+	ret = sec_do_cipher_task(c_ctx->qp, c_ctx);
+	if (ret != 0)
+		goto free_buffer;
+
+	ret = sec_update_iv(c_ctx);
+	if (ret != 0) {
+		EMSG("sec_update_iv failed. ret = 0x%"PRIx32, ret);
+		goto free_buffer;
+	}
+
+	memcpy(dupdate->dst.data,
+	       c_ctx->out + padding_size, dupdate->src.length);
+
+free_buffer:
+	sec_free_buffer(c_ctx);
+	return ret;
+}
+
+/*
+ * Finalize of the cipher operation
+ *
+ * @ctx	Caller context variable or NULL
+ */
+static void sec_cipher_final(void *ctx __unused)
+{
+}
+
+static void sec_cipher_copy_state(void *dst_ctx, void *src_ctx)
+{
+	struct sec_cipher_ctx *dst_c_ctx = (struct sec_cipher_ctx *)dst_ctx;
+	struct sec_cipher_ctx *src_c_ctx = (struct sec_cipher_ctx *)src_ctx;
+
+	dst_c_ctx->alg = src_c_ctx->alg;
+	dst_c_ctx->mode = src_c_ctx->mode;
+	dst_c_ctx->encrypt = src_c_ctx->encrypt;
+	dst_c_ctx->offs = src_c_ctx->offs;
+
+	if (src_c_ctx->key_len) {
+		dst_c_ctx->key_len = src_c_ctx->key_len;
+		dst_c_ctx->c_key_len = src_c_ctx->c_key_len;
+		memcpy(dst_c_ctx->key, src_c_ctx->key, dst_c_ctx->key_len);
+		dst_c_ctx->key_dma = virt_to_phys(dst_c_ctx->key);
+		assert(dst_c_ctx->key_dma);
+	}
+
+	if (src_c_ctx->iv_len) {
+		dst_c_ctx->iv_len = src_c_ctx->iv_len;
+		memcpy(dst_c_ctx->iv, src_c_ctx->iv, dst_c_ctx->iv_len);
+		dst_c_ctx->iv_dma = virt_to_phys(dst_c_ctx->iv);
+		assert(dst_c_ctx->iv_dma);
+	}
+}
+
+/*
+ * Registration of the Cipher Driver
+ */
+static struct drvcrypt_cipher driver_cipher = {
```

Comment:
`const`?

Reply:
I have tried, but it doesn't seem to be useful.
```c
core/drivers/crypto/hisilicon/sec/sec_cipher.c: In function ‘sec_init_cipher’:
core/drivers/crypto/hisilicon/sec/sec_cipher.c:767:44: warning: passing argument 1 of ‘drvcrypt_register_cipher’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  767 |  TEE_Result ret = drvcrypt_register_cipher(&driver_cipher);
      |                                            ^~~~~~~~~~~~~~
In file included from core/drivers/crypto/hisilicon/sec/../include/hisi_cipher.h:9,

```